addons/l10n_id_efaktur_coretax/tests/test_l10n_id_efaktur_coretax.py

```diff
@@ -0,0 +1,465 @@
+            ],
+        })
+        out_invoice.action_post()
+
+        with self.assertRaises(ValidationError) as val:
+            out_invoice.download_efaktur()
+        self._verify_error_message(val, 2, ["NPWP for customer", "is not taxable"])
+
+        # activate PKP, fill in VAT, change document type to passport
+        partner.vat = "1234567890123478"
+        partner.l10n_id_pkp = True
+        partner.l10n_id_buyer_document_type = 'Passport'
+
+        with self.assertRaises(ValidationError) as val:
+            out_invoice.download_efaktur()
+        self._verify_error_message(val, 1, ["Document number for customer"])
+
+    def test_efaktur_invalid_kode_07_08(self):
+        """ Test to extra fields are filled in when code 07 or 08 is used """
+        out_invoice = self.env['account.move'].create({
+            'move_type': 'out_invoice',
+            'partner_id': self.partner_id.id,
+            'invoice_date': '2019-05-01',
+            'date': '2019-05-01',
+            'invoice_line_ids': [
+                (0, 0, {'name': 'line1', 'price_unit': 100000, 'quantity': 1})
+            ],
+            'l10n_id_kode_transaksi': '07'
+        })
+
+        out_invoice.action_post()
+
+        with self.assertRaises(ValidationError) as val:
+            out_invoice.download_efaktur()
+        self._verify_error_message(val, 1, "(Kode 07)")
+
+        out_invoice_2 = self.env['account.move'].create({
+            'move_type': 'out_invoice',
+            'partner_id': self.partner_id.id,
+            'invoice_date': '2019-05-01',
+            'date': '2019-05-01',
+            'invoice_line_ids': [
+                (0, 0, {'name': 'line1', 'price_unit': 100000, 'quantity': 1})
+            ],
+            'l10n_id_kode_transaksi': '08'
+        })
+        out_invoice_2.action_post()
+
+        with self.assertRaises(ValidationError) as val:
+            out_invoice_2.download_efaktur()
+        self._verify_error_message(val, 1, "(Kode 08)")
+
+    # ============================
+    # Test eFaktur XML content
+    # ============================
+
+    def test_efaktur_xml_partner_edit(self):
+        """ Test the effect of changing customer information/fields towards the generated XML content"""
+        self.partner_id.write({
+            "vat": "1234567890999999",
+            "l10n_id_tku": "222222",
+            "l10n_id_buyer_document_type": "Passport",
+            "l10n_id_buyer_document_number": "A123456",
+            "l10n_id_pkp": True
+        })
+
+        out_invoice = self.env["account.move"].create({
+            'move_type': 'out_invoice',
+            'partner_id': self.partner_id.id,
+            'invoice_date': '2019-05-01',
+            'date': '2019-05-01',
+            'invoice_line_ids': [
+                (0, 0, {'product_id': self.product_id.id, 'name': 'line1', 'price_unit': 100000, 'quantity': 1})
+            ],
+            'l10n_id_kode_transaksi': '04'
```

Comment:
```suggestion
            'l10n_id_kode_transaksi': '04',
```
always add ',' in a dict, even on the last line.
This avoid changing the ownership of that line if we need to add something just for a ,